http/dotnet/httpclient/src/Middleware/ChaosHandler.cs

```diff
@@ -0,0 +1,241 @@
+        /// </summary>
+        /// <param name="chaosHandlerOptions">Optional parameter to change default behavior of handler.</param>
+        public ChaosHandler(ChaosHandlerOption chaosHandlerOptions = null)
+        {
+            _chaosHandlerOptions = chaosHandlerOptions ?? new ChaosHandlerOption();
+            _random = new Random(DateTime.Now.Millisecond);
+            LoadKnownFailures(_chaosHandlerOptions.KnownChaos);
+        }
+
+        /// <summary>
+        /// Sends the request
+        /// </summary>
+        /// <param name="request">The request to send.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> for the request.</param>
+        /// <returns></returns>
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if(request == null)
+                throw new ArgumentNullException(nameof(request));
+
+            // Select global or per request options
+            var chaosHandlerOptions = request.GetMiddlewareOption<ChaosHandlerOption>() ?? _chaosHandlerOptions;
+
+            HttpResponseMessage response = null;
+            // Planned Chaos or Random?
+            if(chaosHandlerOptions.PlannedChaosFactory != null)
+            {
+                response = chaosHandlerOptions.PlannedChaosFactory(request);
+                if(response != null)
+                {
+                    response.RequestMessage = request;
+                    if(_logger.IsEnabled("PlannedChaosResponse"))
+                        _logger.Write("PlannedChaosResponse", response);
+                }
+            }
+            else
+            {
+                if(_random.Next(100) < chaosHandlerOptions.ChaosPercentLevel)
+                {
+                    response = CreateChaosResponse(chaosHandlerOptions.KnownChaos ?? _knownFailures);
+                    response.RequestMessage = request;
+                    if(_logger.IsEnabled("ChaosResponse"))
+                        _logger.Write("ChaosResponse", response);
+                }
+            }
+
+            if(response == null)
+            {
+                response = await base.SendAsync(request, cancellationToken);
+            }
+            return response;
+        }
+
+        private HttpResponseMessage CreateChaosResponse(List<HttpResponseMessage> knownFailures)
+        {
+            var responseIndex = _random.Next(knownFailures.Count);
+            return knownFailures[responseIndex];
+        }
+
+        private void LoadKnownFailures(List<HttpResponseMessage> knownFailures)
+        {
+            if(knownFailures?.Any() ?? false)
+            {
+                _knownFailures = knownFailures;
+            }
+            else
+            {
+                _knownFailures = new List<HttpResponseMessage>
+                {
+                    Create429TooManyRequestsResponse(new TimeSpan(0, 0, 3)),
```

Comment:
I believe the handler should implement IDisposable and dispose the responses if it created them.

Reply:
Resolved via cf2ade9